DllImportGenerator/designs/StructMarshalling.md

```diff
@@ -0,0 +1,127 @@
+# Struct Marshalling
+
+As part of the new source-generated direction for .NET Interop, we are looking at various options for supporting marshaling user-defined struct types.
+
+These types pose an interesting problem for a number of reasons listed below. With a few constraints, I believe we can create a system that will enable users to use their own user-defined types and pass them by-value to native code.
+
+## Problems
+
+- Unmanaged vs Blittable
+  - The C# language (and Roslyn) do not have a concept of "blittable types". It only has the concept of "unmanaged types", which is similar to blittable, but differs for `bool`s and `char`s. `bool` and `char` types are "unmanaged", but are never (in the case of `bool`), or only sometimes (in the case of `char`) blittable. As a result, we cannot use the "is this type unmanaged" check in Roslyn for structures.
+- Limited type information in ref assemblies.
+  - In the ref assemblies generated by dotnet/runtime, we save space and prevent users from relying on private implementation details of structures by emitting limited information about their fields. Structures that have at least one non-object field are given a private `int` field, and structures that have at least one field that transitively contains an object are given one private `object`-typed field. As a result, we do not have full type information at code-generation time for any structures defined in the BCL when compiling a library that uses the ref assemblies.
+- Private reflection
+  - Even when we do have information about all of the fields, we can't emit code that references them if they are private, so we would have to emit unsafe code and calculate offsets manually to support marshaling them.
+  
+## Possible Solution A - Opt-In Structure Interop
+
+We've been working around another problem for a while in the runtime-integrated interop design: The user can use any type that is non-auto layout and has fields that can be marshaled in interop. This has lead to various issues where types that were not intended for interop usage became usable and then we couldn't update their behavior to be special-cased since users may have been relying on the generic behavior (`Span<T>`, `Vector<T>` come to mind).
+
+I propose an opt-in design where the owner of a struct has to explicitly opt-in to usage for interop. This enables our team to add special support as desired for various types such as `Span<T>` while also avoiding the private reflection and limited type information issues mentioned above.
+
+This design would use three attributes:
+
+```csharp
+
+public class GeneratedMarshallingAttribute : Attribute {}
+
+public class BlittableTypeAttribute : Attribute {}
+
+public class NativeStructMarshallingAttribute : Attribute
+{
+     public NativeStructMarshallingAttribute(Type nativeStructType, string managedToNativeMarshalMethod, string nativeToManagedMarshalMethod, string nativeFreeMethod) {}
```

Comment:
Is it worth the complexity to make the method names configurable? Can the method names follow a convention instead?

Reply:
As long as we make the convention well documented so people writing manual marshalling code can easily opt in to this system (and validate that they have it set up correctly via an analyzer), I am totally fine making them follow a convention.

Reply:
I would prefer to have an interface that needs to be implemented here or perhaps some kind of registration of function pointer contract.

>Is it worth the complexity to make the method names configurable?

What complexity do you foresee with this approach?

Reply:
> I would prefer to have an interface that needs to be implemented here or perhaps some kind of registration of function pointer contract.

Both interfaces or function pointers are extra indirections. I think they would just get into way for generating efficient marshalers.

> What complexity do you foresee with this approach?

The configurable names make the sources more verbose, but I do not see the value that they add.

Reply:
The extra indirections are a concern for me as well. I have measured the `ICustomMarshaler` overheard before and am concerned about how we could improve that. I am also conflicted by success of use. I know this is an advanced feature so that argument is fair. The other side of that is longevity that I want to ensure provides some intuitive sense to adapt and grow without becoming an incrementally designed and free form mess (i.e. support matrix of sadness).

Agreed upon convention is fine but not for naming in my experience because it embeds current thinking and doesn't typically allow much room for change. I don't mind the above in the sense the supplied string represents a method to call that has a defined signature and we insert that string into a format: `"{ret} = {USER_SUPPLIED}({args});"`. If it compiles great, otherwise the supplied function needs to be "fixed".

Reply:
The source generator should be able to inspect methods available on NativeType and pick the best one to call for given scenario; or no method at all if it is not needed.

For example: If `{NativeType}.Free` method exists, the source generator can it to perform cleanup and also generate the associated try/finally exception handling. If `{NativeType}.Free` method does not exist, the source generator can skip generating all that.

Reply:
Alright I am becoming less concerned about this. Our support assembly could contain a collection of these "NativeType"s and this would allow interop library authors the flexibility to mix/compose how they want to marshal the managed type. This scales pretty well as we can fold popular "NativeType"s into the support assembly. Is that the path you are seeing here?

Reply:
Yes, this can be one of the paths that this enables.

Reply:
I don't think simply matching by name alone is good enough. Depending on the exact scenario you may have helper methods or other members available on a type that aren't necessarily meant for use in marshaling.

`IAMDevMemoryAllocator` in particular has a method named `Free()`, as do several other types

Reply:
I think it needs to be opt-in in some fashion, likely by matching some attribute + signature to ensure that the correct thing always happens.

Reply:
It is opt-in by having dedicated `NativeType` that contains the marshalling methods. This type is different from your mainstream stream, and so there should not be risk of collisions.